termux-shared/src/main/java/com/termux/shared/logger/Logger.java

```diff
@@ -0,0 +1,330 @@
+    }
+
+    public static String getMessageAndStackTraceString(String message, Throwable throwable) {
+        if (message == null && throwable == null)
+            return null;
+        else if (message != null && throwable != null)
+            return message + ":\n" + getStackTraceString(throwable);
+        else if (throwable == null)
+            return message;
+        else
+            return getStackTraceString(throwable);
+    }
+
+    public static String getMessageAndStackTracesString(String message, List<Throwable> throwableList) {
+        if (message == null && (throwableList == null || throwableList.size() == 0))
+            return null;
+        else if (message != null && (throwableList != null && throwableList.size() != 0))
+            return message + ":\n" + getStackTracesString(null, getStackTraceStringArray(throwableList));
+        else if (throwableList == null || throwableList.size() == 0)
+            return message;
+        else
+            return getStackTracesString(null, getStackTraceStringArray(throwableList));
+    }
+
+    public static String getStackTraceString(Throwable throwable) {
+        if (throwable == null) return null;
+
+        String stackTraceString = null;
+
+        try {
+            StringWriter errors = new StringWriter();
+            PrintWriter pw = new PrintWriter(errors);
+            throwable.printStackTrace(pw);
+            pw.close();
+            stackTraceString = errors.toString();
+            errors.close();
+        } catch (IOException e1) {
+            e1.printStackTrace();
+        }
+
+        return stackTraceString;
+    }
+
+    public static String[] getStackTraceStringArray(Throwable throwable) {
+        return getStackTraceStringArray(Collections.singletonList(throwable));
+    }
+
+    public static String[] getStackTraceStringArray(List<Throwable> throwableList) {
+        if (throwableList == null) return null;
+
+        final String[] stackTraceStringArray = new String[throwableList.size()];
+        for (int i = 0; i < throwableList.size(); i++) {
+            stackTraceStringArray[i] = getStackTraceString(throwableList.get(i));
+        }
+        return stackTraceStringArray;
+    }
+
+    public static String getStackTracesString(String label, String[] stackTraceStringArray) {
+        if (label == null) label = "StackTraces:";
+        StringBuilder stackTracesString = new StringBuilder(label);
+
+        if (stackTraceStringArray == null || stackTraceStringArray.length == 0) {
+            stackTracesString.append(" -");
+        } else {
+            for (int i = 0; i != stackTraceStringArray.length; i++) {
+                if (stackTraceStringArray.length > 1)
+                    stackTracesString.append("\n\nStacktrace ").append(i + 1);
+
+                stackTracesString.append("\n```\n").append(stackTraceStringArray[i]).append("\n```\n");
+            }
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 210:219, 228:237.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.